apps/nccl/src/allreduce.hpp

```diff
@@ -478,13 +485,20 @@ __global__ void __launch_bounds__(32, 1)
     }
     dst[idx] = data;
   }
+  __syncthreads();
+
+  deviceSyncer->sync(gridDim.x);
```

Comment:
Do we need deviceSyncer here? If we run allreduce sequentially, it will be launched in different kernel with same stream. And `Kernels launched into the same stream are guaranteed to execute consecutively`

Reply:
I think kernel launch works like a barrier. Before we launch another kernel, it will make sure the previous barrier finished its' work. Here is the doc for this: https://docs.nvidia.com/cuda/cuda-c-programming-guide/#managing-data-visibility-and-concurrent-cpu-gpu-access-with-streams `Kernels launched into the same stream are guaranteed to execute consecutively,`

Reply:
I am not worried about subsequent kernel launch. What if `block-0` is faster than the other blocks in the same kernel launch?

Reply:
Okay, got you point. If block-0 is faster than the other blocks, it will update the deviceFlag. But since other threadblock still use cached flag, seems there is no issue? Other threadblock will use the latest deviceFlag in next kernel launch